### ISSUE TITLE
False positive with W0102: non-empty dict marked as dangerous default value

### ISSUE THREAD

**Issue by MuriloScarpaSitonio** [PyCQA/pylint#3642]
### Steps to reproduce
1. Create file `t.py`

```
def find(_filter={'_id': 0}):
    print(_filter)
```

2. run `pylint t.py`

### Current behavior
`t.py:1:0: W0102: Dangerous default value {} as argument (dangerous-default-value)`

### Expected behavior
no errors, no warnings

### pylint --version output
`pylint 2.5.2`
`astroid 2.4.1`
`Python 3.7.2 (tags/v3.7.2:9a3ffc0492, Dec 23 2018, 22:20:52) [MSC v.1916 32 bit (Intel)]`

**Comment by ethan-leba** [comment]
Unless I'm misunderstanding, I don't think this is a false positive?
```
Python 3.7.1 (v3.7.1:260ec2c36a, Oct 20 2018, 03:13:28)
[Clang 6.0 (clang-600.0.57)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> def x(y = {1: 2}):
...     print(y)
...     y[3] = 4
...
>>> x()
{1: 2}
>>> x()
{1: 2, 3: 4}
```

**Comment by Arrowana** [comment]
To summarize the problem than @ethan-leba illustrated with some code:
This is dangerous because a mutable default value retains its mutated state on subsequent calls.

As explained here in other words
http://pylint-messages.wikidot.com/messages:w0102

However, we might argue that the warning is misleading since "default value {}" {} is not the dictionary passed as an argument

**Comment by MuriloScarpaSitonio** [comment]
I understand the dangerous of using a mutable default value and I suggest switching the warning message for something like "Dangerous mutable default value as argument". 

However, this is dangerous for all sorts of scenarios? (I know that pylint isn't supposed to check the functionality of my code, just trying to clarify this anti-pattern)

```
>>> def find(_filter={'_id': 0}):
...     print({**find.__defaults__[0], **_filter})
...
>>> find()
{'_id': 0}
>>> find({'a': 1})
{'_id': 0, 'a': 1}
>>> find()
{'_id': 0}
>>> find({'a': 1, 'b': 2})
{'_id': 0, 'a': 1, 'b': 2}
```

**Comment by ethan-leba** [comment]
Here's another common idiom for mutable default values:
```
def find(_filter=None):
    _filter = _filter or {'_id' : 0}
```
I like this pattern b/c it's a one-liner, but the way you wrote it works too.

It's true that it's not actually a problem if you don't mutate the value, but the message is there to warn to you to keep that in mind. Also, the 'dangerous-ness' of mutating a default value doesn't show up until the second time you call the function, so it can make tracking down the bug difficult if you accidentally ending up mutating a default argument.

I think emphasizing the fact that the value is dangerous because it's mutable is a good idea, and maybe grab the name of the class for the message? So your code above would display `Dangerous mutable default value 'dict' as argument`.

As it stands, this error isn't raised on user-defined classes, this might also make a good addition to the check. Maybe even have a type whitelist for default values, to built-in immutable values?
```
class Data:
    def __init__(self):
        self.field = []


def x(data=Data()): <-- no error
    data.field.append(0)
```

**Comment by Pierre-Sassoulas** [comment]
As @ethan-leba said this is working as intended. You can disable the check if you want to use this amazing python feature. (My perl loving friends make fun of me because of those mutable default arguments, now I'll be able to point them to someone who actually like the feature 😄)
